docs/proposals/ktl.md

```diff
@@ -0,0 +1,243 @@
+
+- creating & updating objects defined by commandline args
+- managing objects using configuration files
+- debugging objects in a cluster
+
+Benefits of this approach:
+
+- having a single binary facilitates discovery for the suite of Kubernetes cli commands
+- static binary makes it simple to distribute the cli across all OS platforms
+- static binary makes it simple for users to install the cli
+- single binary limits the size of the install (MB) vs distributing ~50 separate go commands (GB)
+- centralizing the development facilitates the construction of shared libraries and infrastructure
+- leveraging the kubernetes/kubernetes repo infrastructure provides a process to build and release
+
+Challenges of this approach:
+
+- the cli cannot be released at a different intervals from the cluster itself
+  - slows down velocity
+  - slows down feedback on alpha and beta features - longer time to GA
+  - each release is heavier weight due to its length
+  - obfuscates version skew as a first class problem
+- individual commands cannot be released at different intervals from one another
+  - some commands should be released every 3 months with the kubernetes cluster
+  - others could be released daily
+- shared cli infrastructure cannot easily be publish to be used by repos / commands outside kubernetes/kubernetes
+  - requires vendoring kubernetes/kubernetes which is painful
+- cli cannnot be owned and maintained independent from unrelated components
+  - submit / merge queue blocks on unrelated tests
+  - GitHub permissions apply to whole repo - cannot add collaborators or maintainers for just the cli code
+  - GitHub notifications for kubernetes/kubernetes are a firehose
+  - hard to manage PRs and issues because they are not scoped to CLI
+
+To address these challenges, sig-cli maintainers have worked toward moving kubectl out of the main repo
+for the past year.  This process has been slow and challenging due to the following reasons:
+
+- many kubectl commands depend on internal kubernetes/kubernetes libraries
+- most commands should not have these dependencies, but removing them requires large rewrites of the commands
+- *some* commands do actually need these dependencies (e.g. convert)
+- continuing to develop in the kubernetes/kubernetes repo results in more *bad* dependencies being added
+  even as we try to remove old ones
+- many commands depend on test infrastructure bespoke to kubernetes/kubernetes, which would need to be moved as well
+
+Additionally, many kubectl commands are laden with technical debt, using anti-patterns for working with APIs
+that do not work with version skew or support extensibility.  We have since grown out of using these patterns,
+but they are still pervasive.  Frequently, is it much faster and effective to rewrite large pieces
+instead of trying to refactor them into different designs.
+
+Some examples of anti-patterns:
+
+- Relying on specific compiled in API types for generic operations (scale, rollback)
```

Comment:
`scale` is will be generic in a month of so.  https://github.com/kubernetes/kubernetes/pull/56077 already merged and we're ready to start switching individual scalers over.  If we're willing to crack a few eggs on the way through it should land without too much trouble.

Reply:
`reconcile` here https://github.com/kubernetes/kubernetes/pull/56197 is also a painful one I think.

Reply:
Glad to hear it.  Sounds like a good start.